modules/renter.go

```diff
@@ -50,12 +54,92 @@ type Allowance struct {
-	Received    uint64    `json:"received"`
-	StartTime   time.Time `json:"starttime"`
-	Error       string    `json:"error"`
+	SiaPath     string         `json:"siapath"`
+	Destination DownloadWriter `json:"destination"`
+	Filesize    uint64         `json:"filesize"`
+	Received    uint64         `json:"received"`
+	StartTime   time.Time      `json:"starttime"`
+	Error       string         `json:"error"`
+}
+
+// DownloadWriter provides an interface which all output writers have to implement.
+type DownloadWriter interface {
+	WriteAt(b []byte, off int64) (int, error)
+}
+
+// DownloadFileWriter is a file-backed implementation of DownloadWriter.
+type DownloadFileWriter struct {
+	f        *os.File
+	Location string
+	offset   uint64
+}
+
+// NewDownloadFileWriter creates a new instance of a DownloadWriter backed by the file named.
+func NewDownloadFileWriter(fname string, offset, length uint64) *DownloadFileWriter {
+	l, _ := os.OpenFile(fname, os.O_CREATE|os.O_WRONLY, defaultFilePerm)
+	return &DownloadFileWriter{
+		f:        l,
+		Location: fname,
+		offset:   offset,
+	}
+}
+
+// WriteAt writes the passed bytes at the specified offset.
+func (dw *DownloadFileWriter) WriteAt(b []byte, off int64) (int, error) {
+	fileOffset := off - int64(dw.offset)
+
+	r, err := dw.f.WriteAt(b, fileOffset)
+	if err != nil {
+		build.ExtendErr("unable to write to download destination", err)
+	}
+	dw.f.Sync()
+
+	return r, err
+}
+
+// DownloadHttpWriter is a http response writer-backed implementation of DownloadWriter.
+// The writer writes all content that is written to the current `offset` directly to the ResponseWriter,
+// and buffers all content that is written at other offsets.
+// After every write to the ResponseWriter the `offset` and `length` fields are updated, and buffer content written until
+type DownloadHttpWriter struct {
+	w              http.ResponseWriter
+	offset         int         // The index in the original file of the last byte written to the response writer.
+	firstByteIndex int         // The index of the first byte in the original file.
+	length         int         // The total size of the slice to be written.
+	buffer         []byte      // Buffer used for storing the chunks until download finished.
+	offsets        map[int]int // Map from offset to length of chunk.
+}
+
+// NewDownloadHttpWriter creates a new instance of http.ResponseWriter backed DownloadWriter.
+func NewDownloadHttpWriter(w http.ResponseWriter, offset, length uint64) *DownloadHttpWriter {
+	return &DownloadHttpWriter{
+		w:              w,
+		offset:         int(offset),
+		firstByteIndex: int(offset),
+		length:         int(length),
+		buffer:         make([]byte, length),
+		offsets:        make(map[int]int),
+	}
+}
+
+// WriteAt buffers parts of the file until the entire file can be
+// flushed to the client. Returns the number of bytes written or an error.
+func (dw *DownloadHttpWriter) WriteAt(b []byte, off int64) (int, error) {
```

Comment:
Yeah, basically to convert the current renter from parallel -> stream, the only thing you need to do is send the finished chunk to a merger instead of to a file. The merger will block until the finished chunk is written to the stream, and it won't write the finished chunk to the stream until all previous chunks are written.

You may want to add some logic that more carefully selects how slow flights of hosts perform downloads, but it would be fairly easy to set a flag such that slow hosts ignore streaming chunk downloads.

Overall, I  think 90% of the optimizations are going to apply to both cases, and it'd be better to have the code heavily overlap with some flags rather than have two separate, but nearly identical, and both highly complex, sets of code to manage the downloads.